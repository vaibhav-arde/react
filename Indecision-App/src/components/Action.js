import React from 'react';


const Action = (props) => {
    return (
        <div>
            <button
                onClick={props.pickAOption}
                disabled={!props.hasOptions}
            >What should I do?</button>
        </div>
    );
}
// class Action extends React.Component {
//     // btnClick() {
//     //     alert(`'What should I do' btn clicked.`)
//     // }
//     render() {
//         return (
//             <div>
//                 <button
//                     onClick={this.props.pickAOption}
//                     disabled={!this.props.hasOptions}
//                 >What should I do?</button>
//             </div>
//         );
//     }
// }

export default Action;