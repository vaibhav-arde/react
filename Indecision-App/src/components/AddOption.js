import React from 'react'

export default class AddOption extends React.Component {
    state = {
        error: undefined
    };

    submitForm = (e) => {
        e.preventDefault();
        console.log('Form Submitted');

        const option = e.target.elements.option.value.trim();
        // console.log(option);

        // option ? alert(`Option submitted is : ${option}`) : alert(`Option can not be empty`)

        const error = this.props.addNewOption(option)
        console.log(`Error is ${error}`)

        this.setState(() => ({ error }));
        // this.setState(() => {
        //     return { error }
        // });

        !error && (e.target.elements.option.value = '')
        // if (!error) {
        //     e.target.elements.option.value = ''
        // }

    }
    render() {
        return (
            <div>
                {this.state.error && <p className='add-option-error'>{this.state.error}</p>}
                <form className='add-option' onSubmit={this.submitForm}>
                    <input className='add-option__input' type='text' name='option'></input>
                    <button className='button'>Add Option</button>
                </form>
            </div>
        );
    }
}