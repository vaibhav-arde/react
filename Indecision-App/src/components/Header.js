import React from 'react';

const Header = (props) => (
    <div>
        <h1>{props.title}</h1>
        {props.subTitle && <h2>{props.subTitle}</h2>}
    </div>
);
Header.defaultProps = {
    title: 'Indecision App!!'
}
// class Header extends React.Component {
//     render() {
//         // console.log(this.props)
//         return (
//             <div>
//                 <h1>{this.props.title}</h1>
//                 <h2>{this.props.subTitle}</h2>
//             </div>
//         );
//     }
// }

export default Header;