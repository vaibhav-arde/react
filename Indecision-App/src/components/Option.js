import React from 'react';

const Option = (props) => (
    <div>
        {props.optionText}
        <button
            className='button button--link'
            onClick={(e) => {
                props.removeSingleBtn(props.optionText)
            }}
        >
            remove
            </button>
    </div>
);
// class Option extends React.Component {
//     render() {
//         return (
//             <div>
//                 {this.props.optionText}
//             </div>
//         );
//     }
// }

export default Option;