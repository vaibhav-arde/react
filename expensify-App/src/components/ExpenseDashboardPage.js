import React from 'react';
import ExpenseList from './ExpensesList';

const ExpenseDashboardPage = () => (
    <div>
        <p>This is from my Dashboard component</p>
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;