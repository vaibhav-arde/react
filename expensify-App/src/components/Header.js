import React from 'react';
import { Link, NavLink } from 'react-router-dom'

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <h3><Link to='/' > Dashboard</Link></h3>
        <h3><Link to='/create' > Create Expense</Link></h3>
        <h3><Link to='/edit' > Edit Expense</Link></h3>
        <h3><Link to='/help' > Get Help</Link></h3>

        <NavLink to='/' activeClassName='is-active' exact={true}> Dashboard</NavLink>
        <NavLink to='/create' activeClassName='is-active' > Create Expense</NavLink>
        <NavLink to='/edit' activeClassName='is-active' > Edit Expense</NavLink>
        <NavLink to='/help' activeClassName='is-active' > Get Help</NavLink>
    </header>
)

export default Header;