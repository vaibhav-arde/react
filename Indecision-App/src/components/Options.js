import React from 'react';
import Option from './Option';


const Options = (props) => (
    <div>
        <button onClick={props.removeAllBtn}>Remove All Options</button>
        <p>Total options available are {props.options.length}</p>
        {props.options.length === 0 && <p>Please add an option to get started!!</p>}
        {props.options.map(option => (
            <Option
                key={option}
                optionText={option}
                removeSingleBtn={props.removeSingleBtn}
            />
        ))}
    </div>
);
// class Options extends React.Component {
//     render() {
//         return (
//             <div>
//                 <button onClick={this.props.removeAllBtn}>Remove All Options</button>
//                 <p>Total options available are {this.props.options.length}</p>
//                 {this.props.options.map(option => <Option key={option} optionText={option} />)}
//             </div>
//         );
//     }
// }
export default Options;