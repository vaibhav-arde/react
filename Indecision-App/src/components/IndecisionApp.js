import React from 'react';
import AddOption from './AddOption';
import Header from './Header';
import Action from './Action';
import Options from './Options';


class IndecisionApp extends React.Component {
    constructor(props) {
        super(props);
        this.removeAllBtn = this.removeAllBtn.bind(this)
        this.pickAOption = this.pickAOption.bind(this)
        this.addNewOption = this.addNewOption.bind(this)
        this.removeSingleBtn = this.removeSingleBtn.bind(this)
        this.state = {
            subTitle: 'Be responsible for all happenings',
            options: []
        }
    }

    componentDidMount() {
        try {
            const json = localStorage.getItem('options');
            const options = JSON.parse(json);
            if (options) {
                this.setState(() => ({ options }));
            }
        } catch (e) {
            // Do nothing at all
        }
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevState.options.length !== this.state.options.length) {
            const json = JSON.stringify(this.state.options);
            localStorage.setItem('options', json);
        }
    }

    componentWillUnmount() {
        console.log(`***componentWillUnmount***`)
    }

    removeAllBtn() {
        this.setState(() => ({ options: [] }));
        // this.setState(() => {
        //     return {
        //         options: []
        //     };
        // })
        console.log(`removeAllBtn is clicked`)
    }

    removeSingleBtn(optionToRemove) {
        console.log(`Single btn is removed`, optionToRemove)
        this.setState((prevState) => ({
            options: prevState.options.filter((option) => optionToRemove !== option)
        }))
    }

    pickAOption() {
        // alert('Test');
        const randomNum = Math.floor(Math.random() * this.state.options.length);
        alert(this.state.options[randomNum])

    }

    addNewOption(option) {
        if (!option) {
            return 'Enter valid value to add item';
        } else if (this.state.options.indexOf(option) > -1) {
            return 'This option already exist';
        }

        this.setState((prevState) => ({ options: prevState.options.concat(option) }));
        // this.setState((prevState) => {
        //     return {
        //         options: prevState.options.concat(option)
        //     }
        // })
    }

    render() {
        return (
            <div>
                <Header title={this.state.title} subTitle={this.state.subTitle} />
                <Action
                    hasOptions={this.state.options.length > 0}
                    pickAOption={this.pickAOption}
                />
                <Options
                    options={this.state.options}
                    removeAllBtn={this.removeAllBtn}
                    removeSingleBtn={this.removeSingleBtn}
                />
                <AddOption
                    addNewOption={this.addNewOption}
                />
            </div>
        );
    }
}

export default IndecisionApp;