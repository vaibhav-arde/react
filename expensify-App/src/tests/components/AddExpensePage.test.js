import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let onSubmit, history, wrapper;

beforeEach(()=>{
    onSubmit = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history} />)
});

test('Should render AddExpensePage correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
})

test('Should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[1]);
})