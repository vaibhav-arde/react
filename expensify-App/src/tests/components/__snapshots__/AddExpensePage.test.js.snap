// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddExpensePage correctly 1`] = `
<div>
  <h1>
    This is from my Add expense component
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
